sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,732 @@
+    dplus = [0 for i in range(N+2)]
+    # d_N = sqrt(a_{2*N})
+    dplus[N] = sqrt(ser[2*N])
+
+    # Use the recurrence relations to find
+    # the value of d_s
+    for s in range(N-1, -2, -1):
+        sm = 0
+        for j in range(s+1, N):
+            sm += dplus[j]*dplus[N+s-j]
+        if s != -1:
+            dplus[s] = (ser[N+s] - sm)/(2*dplus[N])
+
+    # Coefficients for the case of d_N = -sqrt(a_{2*N})
+    dminus = [-x for x in dplus]
+
+    # The third equation in Eq 5.15 of the thesis is WRONG!
+    # d_N must be replaced with N*d_N in that equation.
+    dplus[-1] = (ser[N+s] - N*dplus[N] - sm)/(2*dplus[N])
+    dminus[-1] = (ser[N+s] - N*dminus[N] - sm)/(2*dminus[N])
+
+    return [dplus, dminus]
+
+
+def construct_d_case_5(ser):
+    # List to store coefficients for plus case
+    dplus = [0, 0]
+
+    # d_0  = sqrt(a_0)
+    dplus[0] = sqrt(ser[0])
+
+    # d_(-1) = a_(-1)/(2*d_0)
+    dplus[-1] = ser[-1]/(2*dplus[0])
+
+    # Coefficients for the minus case are just the negative
+    # of the coefficients for the positive case.
+    dminus = [-x for x in dplus]
+
+    return [dplus, dminus]
+
+
+def construct_d_case_6(num, den, x):
+    # s_oo = lim x->0 1/x**2 * a(1/x) which is equivalent to
+    # s_oo = lim x->oo x**2 * a(x)
+    s_inf = limit_at_inf(Poly(x**2, x)*num, den, x)
+
+    # d_(-1) = (1 +- sqrt(1 + 4*s_oo))/2
+    return [[(1 + sqrt(1 + 4*s_inf))/2], [(1 - sqrt(1 + 4*s_inf))/2]]
+
+
+def construct_d(num, den, x, val_inf):
+    N = -val_inf//2
+    # Multiplicity of oo as a pole
+    mul = -val_inf if val_inf < 0 else 0
+    ser = rational_laurent_series(num, den, x, oo, mul, 1)
+
+    # Case 4
+    if val_inf < 0:
+        d = construct_d_case_4(ser, N)
+
+    # Case 5
+    elif val_inf == 0:
+        d = construct_d_case_5(ser)
+
+    # Case 6
+    else:
+        d = construct_d_case_6(num, den, x)
+
+    return d
+
+
+def rational_laurent_series(num, den, x, x0, mul, n):
+    # Basic Idea for this function is explained here:
+    # https://github.com/sympy/sympy/pull/21459#issuecomment-850500922
+
+    # Variable for symbolic index
+    m = symbols('m')
+    one = Poly(1, x, extension=True)
+
+    if x0 == oo:
+        # Series at x = oo is equal to first transforming
+        # the function from x -> 1/x and finding the
+        # series at x = 0
+        num, den = inverse_transform_poly(num, den, x)
+        x0 = S(0)
+
+    if x0:
+        # For an expansion about a non-zero point, a
+        # transformation from x -> x + x0 must be made
+        num = num.transform(Poly(x + x0, x, extension=True), one)
+        den = den.transform(Poly(x + x0, x, extension=True), one)
+        num, den = num.cancel(den, include=True)
+
+    # Remove the pole from the denominator if the series
+    # expansion is about one of the poles
+    num, den = (num*x**mul).cancel(den, include=True)
+
+    # IndexedBase object to represent recurrence relation
+    # for the Laurent series coefficients
+    c = IndexedBase('c')
```

Comment:
I don't really understand why IndexedBase is being used here. It seems like a complicated way to get from the denominator to the recursion relation. Is the recursion relation not just a linear equation having the same coefficients as the denominator polynomial?

Reply:
Yes, but I am using `IndexedBase` because it is easy to represent the recurrence relation and it allows symbolic indexing. In one of the commits above, I tried removing it and adding an approach without IndexedBase, but it is hard to construct the coefficients since I have to store them in a list and it doesn't have symbolic indexing.

Reply:
I don't see why symbolic indexing is needed:
```python
from sympy.polys.polytools import parallel_poly_from_expr


def ratfunc_series_at_pole(f, x, r, m, n):
    """Series of f(x) around pole r of order m up to nth term"""

    # Shift so we can compute series around x=0
    fr = f.subs(x, r + x) * x**m
    num, den = fr.as_numer_denom()
    (num, den), opt = parallel_poly_from_expr((num, den), x)

    # Equate coefficients for the first terms:
    maxdegree = 1 + max(num.degree(), den.degree())
    syms = symbols(f'a:{maxdegree}', cls=Dummy)
    diff = num - den * Poly(syms[::-1], x)
    coeff_diffs = diff.all_coeffs()[::-1][:maxdegree]
    (coeffs,) = linsolve(coeff_diffs, syms)

    # Use the recursion relation for the rest:
    recursion = den.all_coeffs()[::-1]
    div, rec_rhs = recursion[0], recursion[1:]
    series = list(coeffs)
    while len(series) < n:
        next_coeff = sum(c*series[-1-n] for n, c in enumerate(rec_rhs)) / div
        series.append(-next_coeff)

    return series[:n]


def series_at_pole(f, x, r, m, n):
    fr = f.subs(x, r + x) * x**m
    return Poly(fr.series(x, n=n).removeO(), x).all_coeffs()[::-1]


f = (1 + 2*x + 3*x**2) / ((x - 1)*(x - 2)*(x - 3)**2)
n = 13
for r, m in roots(denom(f)).items():
    series1 = ratfunc_series_at_pole(f, x, r, m, n)
    series2 = series_at_pole(f, x, r, m, n)
    assert series1 == series2
```